### PR TITLE
tests: use journalctl cursors instead rotating logs

### PR DESCRIPTION
**Pull request by sergiocazzolato**
This change is intended to use journalctl cursors instead of rotating log files, and also to use a single implementation for all the systems (currently we have a different implementation for ubuntu 14.04).

This also deals with the issue generated with journalctl --sync where the
command hungs.

### CODE REVIEW

**Comment by bboozzoo** [comment]
```
2018-05-11 06:07:14 Error executing google:ubuntu-18.04-64:tests/main/snapd-reexec : 
-----
+ '[' '' = 0 ']'
+ echo 'Ensure we re-exec by default'
Ensure we re-exec by default
+ MATCH 'DEBUG: restarting into'
+ /usr/bin/env SNAPD_DEBUG=1 snap list
+ . /home/gopath/src/github.com/snapcore/snapd/tests/lib/dirs.sh
++ export SNAP_MOUNT_DIR=/snap
++ SNAP_MOUNT_DIR=/snap
++ export LIBEXECDIR=/usr/lib
++ LIBEXECDIR=/usr/lib
++ export MEDIA_DIR=/media
++ MEDIA_DIR=/media
++ case "$SPREAD_SYSTEM" in
+ . /home/gopath/src/github.com/snapcore/snapd/tests/lib/journalctl.sh
++ JORNALCTL_CURSOR_FILE=/home/gopath/src/github.com/snapcore/snapd/journalctl_cursor
+ echo 'Ensure that we do not re-exec into older versions'
Ensure that we do not re-exec into older versions
+ systemctl stop snapd.service snapd.socket
+ echo 'mount something older than our freshly build snapd'
mount something older than our freshly build snapd
+ echo VERSION=1.0
+ mount --bind /tmp/old-info /snap/core/current/usr/lib/snapd/info
+ systemctl start snapd.service snapd.socket
+ snap list
Name  Version                   Rev   Tracking  Developer  Notes
core  16-2.32.6+git716.83cec93  4641  edge      canonical  core
+ MATCH 'core snap \(at .*\) is older \(.*\) than distribution package'
+ get_journalctl_log
++ cat /home/gopath/src/github.com/snapcore/snapd/journalctl_cursor
+ cursor='Binary file (standard input) matches'
+ journalctl --cursor 'Binary file (standard input) matches'
Failed to seek to cursor: Invalid argument
error: pattern not found, got:
-----
```

File: tests/lib/journalctl.sh [hunk 1]
```diff
@@ -0,0 +1,30 @@
+#!/bin/sh
+
+JORNALCTL_CURSOR_FILE="${SPREAD_PATH}"/journalctl_cursor
+
+get_last_journalctl_cursor(){
+    # It is not being used journalctl json output because jq is not installed on
+    # core systems
+    cursor=$(journalctl --output=export -n1 | grep -e '__CURSOR=')
+    echo ${cursor#*=}
```
Comment (bboozzoo) [inline tests/lib/journalctl.sh]: `${cursor#__CURSOR=}` is probably safer and more explicit

File: tests/lib/journalctl.sh [hunk 2]
```diff
@@ -0,0 +1,30 @@
+#!/bin/sh
+
+JORNALCTL_CURSOR_FILE="${SPREAD_PATH}"/journalctl_cursor
```
Comment (bboozzoo) [inline tests/lib/journalctl.sh]: `JOURNALCTL_CURSOR_FILE`

File: tests/main/refresh-delta/task.yaml [hunk 3]
```diff
@@ -19,8 +19,10 @@ prepare: |
     snap install --edge $SNAP_NAME
 
 execute: |
+    . "$TESTSLIB/journalctl.sh"
```
Comment (bboozzoo) [inline tests/main/refresh-delta/task.yaml]: Can you include this in every test that sources journalctl.sh?
```
#shellcheck source=tests/lib/journalctl.sh
```

File: tests/lib/journalctl.sh [hunk 4]
```diff
@@ -0,0 +1,34 @@
+#!/bin/sh
+
+JOURNALCTL_CURSOR_FILE="${SPREAD_PATH}"/journalctl_cursor
+
+get_last_journalctl_cursor(){
+    # It is not being used journalctl json output because jq is not installed on
+    # core systems
+    if [ -z "$(which jq)" ]; then
+        cursor=$(journalctl --output=export -n1 | grep -e '__CURSOR=')
```
[the remark this answers is not in the document]
Comment (zyga) [reply tests/lib/journalctl.sh]: Hmmm, can you expand on that statement? What did you mean?

File: tests/lib/journalctl.sh [hunk 5]
```diff
@@ -0,0 +1,34 @@
+#!/bin/sh
+
+JOURNALCTL_CURSOR_FILE="${SPREAD_PATH}"/journalctl_cursor
+
+get_last_journalctl_cursor(){
+    # It is not being used journalctl json output because jq is not installed on
+    # core systems
+    if [ -z "$(which jq)" ]; then
```
Comment (zyga) [inline tests/lib/journalctl.sh]: I would recommend that we have one code path for all cases. The jq-free version looks fine so let's just use it all the time.

File: tests/main/snap-userd-desktop-app-autostart/task.yaml [hunk 6]
```diff
@@ -29,12 +30,10 @@ execute: |
     test -e ~/snap/test-snapd-xdg-autostart/current/foo-autostarted
 
     if [[ "$SPREAD_SYSTEM" != ubuntu-14.04-* ]]; then
-        journalctl --flush
-        test "$(journalctl -t foo.desktop --after-cursor=$cursor | wc -l)" -eq 1
+        test "$(journalctl -t foo.desktop --cursor "$cursor" | wc -l)" -gt 0
```
Comment (zyga) [inline tests/main/snap-userd-desktop-app-autostart/task.yaml]: You have a function for this, no? Perhaps the function is too trivial and should be removed.

File: tests/main/snap-userd-desktop-app-autostart/task.yaml [hunk 7]
```diff
@@ -29,12 +30,10 @@ execute: |
...
     else
-        # 14.04 journalctl does not have cursor, neither --identifier support,
-        # nor --flush
-        test "$(journalctl | grep foo.desktop | wc -l)" -eq 1
+        # 14.04 journalctl does not have cursor, neither --identifier support
```
Comment (zyga) [inline tests/main/snap-userd-desktop-app-autostart/task.yaml]: Wait, so does 14.04 have or doesn't have cursors? If it doesn't have cursors then what are we doing in all the generic function and test cases above?

Comment (bboozzoo) [reply tests/main/snap-userd-desktop-app-autostart/task.yaml]: IIRC @sergiocazzolato mentioned that systemd got a SRU and it currently has cursors. The comment comes from the time I added the test (and there was no cursors support then).

File: tests/lib/journalctl.sh (hunk 4 above)
[the remark this answers is not in the document]
Comment (bboozzoo) [reply tests/lib/journalctl.sh]: `--output=export` is for use with systemd-journal-remote,  and it's not usable with shell's eval like `dbus-launch` for instance is.

File: tests/main/snap-userd-desktop-app-autostart/task.yaml (hunk 7 above)
Comment (zyga) [reply tests/main/snap-userd-desktop-app-autostart/task.yaml]: Ah, this explains everything. In that case, please update the comment.

File: tests/main/snap-userd-desktop-app-autostart/task.yaml (hunk 6 above)
Comment (sergiocazzolato) [reply tests/main/snap-userd-desktop-app-autostart/task.yaml]: The function that we already have checks against the predefined cursor, in this case we are checking agains another cursor. I'll update the function to reuse it.

Comment (zyga) [reply tests/main/snap-userd-desktop-app-autostart/task.yaml]: I think we had a function that took the cursor as argument but perhaps I was mistaken,

**Pull request merged**